classes/cron_manager.php

```diff
@@ -0,0 +1,115 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_excimer;
+
+defined('MOODLE_INTERNAL') || die();
+
+/**
+ * Manages the profiling of cron runs.
+ *
+ * @package   tool_excimer
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
+ * @copyright 2021, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class cron_manager {
+    /**
+     * Sets callbacks to handle cron profiling.
+     *
+     * @param \ExcimerProfiler $profiler
+     * @param \ExcimerTimer $timer
+     * @param float $started
+     * @throws \dml_exception
+     */
+    public static function set_callbacks(\ExcimerProfiler $profiler, \ExcimerTimer $timer, float $started): void {
+        $timer->setCallback(function($s) use ($profiler, $started) {
+            $log = $profiler->getLog();
+            cron_manager::process($log, $started);
+        });
+
+        \core_shutdown_manager::register_function(
+            function() use ($profiler, $timer, $started) {
+                $timer->stop();
+                $profiler->stop();
+                $log = $profiler->flush();
+                cron_manager::process($log, $started);
+            }
+        );
+    }
+
+    /**
+     * Process a batch of Excimer logs.
+     *
+     * @param \ExcimerLog $log
+     * @param float $started
+     * @throws \dml_exception
+     */
+    protected static function process(\ExcimerLog $log, float $started): void {
+        $current = microtime(true);
+        $sampleperiod = manager::sample_period();
+        $flamed3node = flamed3_node::from_excimer($log);
+        $nodes = self::extract_task_nodes($flamed3node);
+        foreach ($nodes as $node) {
+            // Duration is between n and n+1 periods, so use n as a simple approximation.
+            $duration = $node->value * $sampleperiod;
+            $request = self::get_request($node);
+            $reasons = manager::get_reasons($request, $duration);
```

Comment:
Not sure best place to comment, but the max number of profiles per request is not honored. I'm happy with this being the same number for web requests, we can split this into its own config later if needed

Reply:
eg from a single cron cli run which has lots of the same task:

![image](https://user-images.githubusercontent.com/187449/148162357-4f3be066-f032-47a9-8ca8-dc2a0af8cdf0.png)

Reply:
I did some debugging runs, and the checking code does get called. The above happens because existing profiles aren't immediately purged if you have a burst situation.

Reply:
This doesn't sound right. At any given point we should only ever persist a new profile if it is already worse than the current ratcheting threshold. So perhaps there is something wrong with the threshold matching logic as this seems to be persisting every single task profile regardless of where.

... ah I see the problem, this all comes back to the root cause of splitting the profiles after the fact rather than resetting the profiler. I can see two other related symptoms: 1) the time started for the profile is the start of the process and not the inferred start of the task. 2) the duration of the task is the complete. It would not surprise me if there are other symptoms. 

Because of point 2, every single task profile is always worse than the previous ones. I'd still prefer it if we just stopped the profiler and reset it each time but I'll leave the internals to you.

Reply:
I have refactored this to work as you suggested. The times are still rough, +/- a sample period, but should be better than previously.